modules/tensor_mechanics/include/userobjects/HomogenizationConstraintIntegral.h

```diff
@@ -0,0 +1,135 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ElementUserObject.h"
+
+// Helpers common to the whole homogenization system
+namespace HomogenizationConstants
+{
+/// Which tensor indices correspond to each constraint number
+//    These are different for large displacements (n^2 constraints)
+//    and small displacements (n(n-1)/2 constraints)
+//    and different depending on the problem dimension
+typedef std::vector<std::pair<unsigned int, unsigned int>> index_list;
+const std::map<bool, std::vector<index_list>> indices{
+    {true,
+     {{{0, 0}},
+      {{0, 0}, {1, 1}, {1, 0}, {0, 1}},
+      {{0, 0}, {1, 0}, {2, 0}, {0, 1}, {1, 1}, {2, 1}, {0, 2}, {1, 2}, {2, 2}}}},
+    {false,
+     {{{0, 0}}, {{0, 0}, {1, 1}, {0, 1}}, {{0, 0}, {1, 1}, {2, 2}, {1, 2}, {0, 2}, {0, 1}}}}};
+
+/// Total number of required constraints for large/small deformations and the
+/// problem dimension
+//    Used to validate user input
+const std::map<bool, std::vector<unsigned int>> required{{true, {1, 4, 9}}, {false, {1, 3, 6}}};
+/// Constraint type: stress/PK stress or strain/deformation gardient
+enum class ConstraintType
+{
+  Stress,
+  Strain
+};
+
+/// Translate from a string to the enum
+inline ConstraintType
+map_string(std::string input)
```

Comment:
we'd use a `MooseEnum` for the input parameter and `getEnum<ConstraintType>()` on the MooseEnum object to get the corresponding enum class value.